Support deflate

lbeschastny opened the pull request:
Most compression middlewares, including [koa-compress], support both `gzip` and `deflate`.

Shouldn't `koa-cash` support it as well?

Though, there are few Cons in supporting it:

 * we'll have to store more data
 * we'll need a little bit more complex compression code to pre-fill all supported encodings while returning only one of them

Any thoughts here?

 [koa-compress]: https://github.com/koajs/compress

lbeschastny commented:
@jonathanong I've played a little bit with compression and adding deflate turned out to be a very simple task: 7e17e26eafe52a463179f6352786b92a381a4858

Though adding an ability to disable cache-level compression still sounds like a very good idea.

lbeschastny commented:
@jonathanong I've thought a bit about caching and compression and came up with the following proposal.

I think we should support deflate, but we should also give users more flexibility by enhancing compression configuration options.

Right now the only configurable option is a `threshold`. I think we should replace it with the following configuration object:

```js
app.use(require('koa-cash')({
  compression: {
    threshold: '1kb',
    gzip: true,
    deflate: true
  }
}))
```

`gzip` and `deflate` options here should control which compressions should be prepared and stored in the cache along with the uncompressed data.

So, the current `koa-cash` behaviour is an equivalent of the following `compression` configuration:

```js
app.use(require('koa-cash')({
  compression: {
    threshold: '1kb',
    gzip: true,
    deflate: false
  }
}))
```

I think we should also let users disable the whole compression thing, so no extra data will be stored in the cache:

```js
app.use(require('koa-cash')({
  compression: false
}))
```

In this case we should avoid setting `Content-Encoding` header and delegate compression negotiation to above middlewares.

----

But I'm not sure what should we do if client will request some unsupported `Content-Encoding`. Right now `koa-cache` serves an uncompressed data whenever it fails to negotiate a compression mechanism, explicitly restricting any compressing middleware to compress the response.

Bearing in mind how fast modern compression libraries are and how important compression may be when serving mobile devices (which grow more and more popular every day), I think this behaviour is not the right one. Even if we don't have any suitable compressed data ready in the cache, it may still be better to serve data compressed, even if it'll mean on-the-fly compression (handled by another middleware).

----

I'll be happy to implement and publish all above changes, but first I want some feedback on my proposal. First, because I would appreciate hearing other opinions on this matter. And second, because I'd like to have some more or less finalized API sketch before coding anything.

niftylettuce commented:
@jonathanong thoughts here?  it's the only PR open here

jonathanong commented:
i don't use this library anyone, but sure I guess?

niftylettuce commented:
Is there an alternative you're using or you just don't use Koa much?

On Mon, May 18, 2020 at 9:00 PM jongleberry <notifications@github.com>
wrote:

> i don't use this library anyone, but sure I guess?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/koajs/cash/pull/42#issuecomment-630530930>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7XBJ7V6AG7OA6QUFHORTRSHR2LANCNFSM4CVPXUIA>
> .
>

jonathanong commented:
i just don't cache responses right now - if I need to, I use a cdn

niftylettuce commented:
Ah, yeah, I ran into so many issues with CDN's and configuration issues eg.
CSP etc that I may be done with them for now.  Perf impact is not much if
your assets are cached already and you have load balancing set up.

On Mon, May 18, 2020 at 11:31 PM jongleberry <notifications@github.com>
wrote:

> i just don't cache responses right now - if I need to, I use a cdn
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/koajs/cash/pull/42#issuecomment-630572891>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7XBL4JK44XCTK3D7K4ILRSIDT5ANCNFSM4CVPXUIA>
> .
>

niftylettuce commented:
@jonathanong can you give me npm owner on this package?

jonathanong commented:
@niftylettuce done

niftylettuce commented:
Thanks @jonathanong check out the v3 rewrite 😄  https://github.com/koajs/cash

niftylettuce commented:
@lbeschastny if you rebase and update your PR (might just be easiest to send a new one in) I will gladly accept it.  I rewrote this package for Koa v3 and rewrote mocha to use ava instead, we also have linting now too and much more.

lbeschastny commented:
> @lbeschastny if you rebase and update your PR (might just be easiest to send a new one in) I will gladly accept it. I rewrote this package for Koa v3 and rewrote mocha to use ava instead, we also have linting now too and much more.

Well, just like @jonathanong I no longer use `koa-cash`.  But I'll gladly rebase my PR if you want to merge it it it's current state.

Though, I still think it may be a good idea to give users an ability to disable build-in compression in favor of more powerful and generic compression solutions like `koa-compress`.

I've already proposed an API in one of my previous comments: https://github.com/koajs/cash/pull/42#issuecomment-260153916

niftylettuce commented:
That would be great, please submit a PR!  We would love to have this.

niftylettuce commented:
Compression by default is disabled in this repository as of v4.0.0

https://github.com/koajs/cash/releases/tag/v4.0.0

If you want insight into how this is actually used in a production app, check out https://github.com/forwardemail/forwardemail.net

lbeschastny commented:
Disabling compression here is actually a good idea. 

But you should probably avoid setting `Content-Encoding` header value to `'identity'` with this behaviour, because it explicitly disables response compression which otherwise could be done on the fly by some compression middleware.


restricting any compressing middleware to compress the response

niftylettuce commented:
Ah, this is the reason why 4.0.0 wasn't working!!! Brilliant.

lbeschastny commented:
This code was meant to prevent other middlewares from compressive already compressed response for a second time.

And since `koa-cash` was always handling compression on its own, there were no additional checks to verify that response caching wasn't skipped due to size threshold or unsupported compression type (which is why this issue was created in the first place).

niftylettuce commented:
v4.0.2 released and deprecation notices added to v4.0.0 and v4.0.1 for the issue

thanks so much @lbeschastny for chiming in here, literally life saver 🙏 👏 🙇 

https://github.com/koajs/cash/releases/tag/v4.0.3

niftylettuce commented:
v4.0.3 released as I saw I missed one instance of identity being set.

Also deprecated v4.0.2 as well just now.

If anyone has further comments/findings lmk.  Thanks again.

https://github.com/koajs/cash/releases/tag/v4.0.4

niftylettuce commented:
Thanks to this package, https://forwardemail.net now scores 100% on Google PageSpeed Insights.

Ref: https://twitter.com/niftylettuce/status/1274684217197121536
